lib/libutils/isoc/bget_malloc.c

```diff
@@ -765,9 +765,9 @@ static void gen_malloc_add_pool(struct malloc_ctx *ctx, void *buf, size_t len)
 	uint32_t exceptions;
 	uintptr_t start = (uintptr_t)buf;
 	uintptr_t end = start + len;
-	const size_t min_len = ((sizeof(struct malloc_pool) + (SizeQuant - 1)) &
-					(~(SizeQuant - 1))) +
-				sizeof(struct bhead) * 2;
+	const size_t min_len = ROUNDUP(sizeof(struct malloc_pool), SizeQuant) +
+			       sizeof(struct bhead) + sizeof(struct bfhead);
+	const size_t init_min_len = min_len + sizeof(struct bfhead);
```

Comment:
This one I can't understand. Are you compiling with `ENABLE_MDBG`, that is, `CFG_TEE_CORE_MALLOC_DEBUG=y` when trying to find the limits?

Reply:
No, I did not enable `ENABLE_MDBG`/`CFG_TEE_CORE_MALLOC_DEBUG`.
I tested calling `gen_malloc_add_pool()` for few small pools. The proposed `min_len` implementation worked for all pools added but the first one that needs this extra size, hence this `init_min_len`.

Note that i'm not that confident in the resulting bget_malloc implementation. Not sure it's bulletproof. In OP-TEE scope, I don't think we expect to add many small pools so I would propose to fix only `min_len` implementation and remove the added `min_len_init` part here.

Reply:
I agree, it's a bit hard to calculate the exactly how much memory the call to `realloc_unlocked()` will need.
How about saying that we'll need at least a 4k page in case it's the first call to `gen_malloc_add_pool()` for this context?

Reply:
4KB is maybe overkilling but I agree an empiric value is better than nothing.

Reply:
Yes, it's for sure too much, but that's the point also. The thing is that the minimum value is quite hard to calculate since you'd need to take what `buf_get_pos()` does into account when predicting the call to `realloc_unlocked()`. For practical purposes it's enough to pick a number that is large enough to pass without any doubt. I suppose we could use 256 bytes instead if you think that's more reasonable.

Reply:
256 byte or 4kB are both fine to me. Okay for 4kB. If a smaller value is needed at some point, we can adjust.
I'll update the P-R with your proposal.

Reply:
Thinking about that, that means TA heap size MUST be higher or equal to 4kB. It is set from `TA_DATA_SIZE` of each TA **user_ta_header_defines.h** header file. We should maybe add a guard like the below in **ta/arch/arm/user_ta_header.c**:
```patch
 /* Keeping the heap in bss */
+#if TA_DATA_SIZE < 4096
+#error TA_DATA_SIZE must be at least 4kByte
+#endif
 uint8_t ta_heap[TA_DATA_SIZE];
 const size_t ta_heap_size = sizeof(ta_heap);
```

Reply:
Makes sense, but it should be with a define from `<malloc.h>` or such.

Reply:
`optee_test/ta/large/include/user_ta_header_defines.h:#define TA_DATA_SIZE	(2 * 1024)` :face_with_head_bandage: 
Fixing OP-TEE test or require 1kB as min heap size?

Reply:
Let's go for 1kB then.

Reply:
updated